include/envoy/router/router.h

```diff
@@ -240,6 +240,16 @@ class RetryState {
   virtual RetryStatus shouldRetryHeaders(const Http::HeaderMap& response_headers,
                                          DoRetryCallback callback) PURE;
 
+  /**
+   * Determines whether given response headers would be retried by the retry policy, assuming
+   * sufficient retry budget and circuit breaker headroom. This is useful in cases where
+   * the information about whether a response is "good" or not is useful, but a retry should
+   * not be attempted for other reasons.
+   * @param response_headers supplies the response headers.
+   * @return bool true if a retry would be warranted based on the retry policy.
+   */
+  virtual bool wouldRetryFromHeaders(const Http::HeaderMap& response_headers) PURE;
```

Comment:
if exposing this isn't desirable, there are some alternatives we could discuss. Right now i'm making sure we never send more than 1 retry for the same upstream request, which is why I needed an API for "i'm not going to retry this, but should I if I was?". The reason I wanted to do that is so we don't send any more retries than in the non-hedging case, the only difference is whether we reset the original request or not.

Reply:
I think exposing this is fine assuming we want the behavior of "accept first response that wouldn't be retried". The other option "accept first response that isn't retried" is simpler but perhaps less ideal as you might be able to wait just a short while for a "better" response.